buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/transform/RestTestTransformer.java

```diff
@@ -0,0 +1,124 @@
+            .filter(transform -> transform instanceof RestTestTransformGlobalSetup)
+            .map(transform -> (RestTestTransformGlobalSetup) transform)
+            .collect(Collectors.toList());
+
+        // Collect any global setup transformations
+        List<RestTestTransformGlobalTeardown> teardownTransforms = transformations.stream()
+            .filter(transform -> transform instanceof RestTestTransformGlobalTeardown)
+            .map(transform -> (RestTestTransformGlobalTeardown) transform)
+            .collect(Collectors.toList());
+
+        // transform the tests and include the global setup and teardown as part of the transform
+        for (ObjectNode test : tests) {
+            Iterator<Map.Entry<String, JsonNode>> testsIterator = test.fields();
+            while (testsIterator.hasNext()) {
+                Map.Entry<String, JsonNode> testObject = testsIterator.next();
+                String testName = testObject.getKey();
+                if ("setup".equals(testName)) {
+                    setupSection = test;
+                }
+                if ("teardown".equals(testName)) {
+                    teardownSection = test;
+                }
+                Map<String, RestTestTransformByObjectKey> objectKeyFinders = transformations.stream()
+                    .filter(transform -> transform instanceof RestTestTransformByObjectKey)
+                    .map(transform -> (RestTestTransformByObjectKey) transform)
+                    .collect(Collectors.toMap(RestTestTransformByObjectKey::getKeyToFind, transform -> transform));
+
+                transformByObjectKeyName(test, objectKeyFinders);
+            }
+        }
+
+        // transform the global setup
+        if (setupTransforms.isEmpty() == false) {
+            int i = 0;
+            for (RestTestTransformGlobalSetup setupTransform : setupTransforms) {
+                ObjectNode result = setupTransform.transformSetup(setupSection);
+                if (result != null && setupSection == null && i++ == 0) {
+                    tests.addFirst(result);
+                }
+            }
+        }
+
+        // transform the global teardown
+        if (setupTransforms.isEmpty() == false) {
+            int i = 0;
+            for (RestTestTransformGlobalTeardown teardownTransform : teardownTransforms) {
+                ObjectNode result = teardownTransform.transformTeardown(teardownSection);
+                if (result != null && teardownSection == null && i++ == 0) {
+                    tests.addLast(result);
+                }
+            }
+        }
+
+        return tests;
+    }
+
+    /**
+     * Recursive method to find any matching object key names.
+     *
+     * @param currentNode      The current node that is being evaluated.
+     * @param objectKeyFinders A Map of the key to find to the
+     */
+    private void transformByObjectKeyName(JsonNode currentNode, Map<String, RestTestTransformByObjectKey> objectKeyFinders) {
```

Comment:
This seems like something that should exist in the transform, as again, we are now coupled to the corresponding interface via an `instanceof` check. Perhaps put this in an abstract implementation of `RestTestTransform` `transformTest` method and have `InjectHeaders` extend that.

Reply:
The `RestTestTransformer` operates against the whole YAML test, and the implementations of `RestTestTransform` operate only on the specific node passed (from one or multiple tests) that it passed. This is to help minimize the number of times we need to traverse the JSON tree. By keeping the JSON tree traversal here (as opposed to the individual transform) allows for multiple transforms with a single traversal for all transforms that are initiated by a named object key.  Currently it is only looks for "do" to inject headers, but future transforms will will be able to implement that interface with a different object key name and do something different like replace the value of object.  Further, this method will likely get extended and renamed in future iterations to allow for more traits that kick off a transform (i.e. replacing one text value with another) while keeping the JSON traversals needed at a minimum.

Reply:
Understood. Does `JsonNode` not contain a reference to it's parent such that the transform itself could make the deterinmation it's "inside" a `do:` block?

Reply:
It does not maintain a parent reference, however, with the recursive approach here it is pretty simple to carry forward the parent (which is only ever an object or an array which kicks off the recursion). 

Not sure that would help however, the goal here is to find all objects with the key "do" once we find that then we can transform it. In this specific case we only need the object itself, in other (future) cases we will need to pass the parent not the thing found. For example if we need an AddSyblingObject then we would need to pass in the parent of found object. 

Perhaps I could change the name of the method to avoid confusion... the method should probably be called `traverseTest`.